pilz_testutils/test/unittest_joint_state_publisher_mock.cpp

```diff
@@ -0,0 +1,147 @@
+/*
+ * Copyright (c) 2021 Pilz GmbH & Co. KG
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <future>
+#include <chrono>
+
+#include <gmock/gmock.h>
+
+#include <sensor_msgs/JointState.h>
+#include <std_msgs/Float64.h>
+#include <pilz_testutils/joint_state_publisher_mock.h>
+#include <pilz_testutils/async_test.h>
+#include <pilz_utils/wait_for_message.h>
+
+namespace pilz_testutils
+{
+class JointStateSubscriberMock
+{
+public:
+  JointStateSubscriberMock()
+  {
+    sub_ = nh_.subscribe<sensor_msgs::JointState>("/joint_states", 1, &JointStateSubscriberMock::cb_mock, this);
+  }
+
+  MOCK_METHOD1(cb_mock, void(const sensor_msgs::JointStateConstPtr&));
+
+private:
+  ros::NodeHandle nh_;
+  ros::Subscriber sub_;
+};
+
+using testing::PrintToString;
+
+static constexpr double COMPUTATION_TOLERANCE{ 0.1 };
+
+MATCHER_P(PositionNear, p, "Position " + std::string(negation ? "is not" : "is") + " near " + PrintToString(p) + ".")
+{
+  return std::abs(arg->position[0] - p) < COMPUTATION_TOLERANCE;
+}
+
+class JointStatePublisherMockTest : public testing::Test, public testing::AsyncTest
+{
+};
+
+TEST_F(JointStatePublisherMockTest, defaultBehaviour)
+{
+  JointStatePublisherMock publisher_mock;
+  publisher_mock.startPublishingAsync();
+
+  pilz_utils::waitForMessage<sensor_msgs::JointState>("/joint_states");
```

Comment:
In tests I would prefer to have a timeout. Else a test failure is hard to understand.